internal/rcm/api.go

```diff
@@ -0,0 +1,148 @@
+// Package rcm provides alternative HTTP API to Weldr.
+// It's primary use case is for the RCM team. As such it is driven solely by their requirements.
+package rcm
+
+import (
+	"encoding/json"
+	"log"
+	"net"
+	"net/http"
+
+	"github.com/google/uuid"
+	"github.com/julienschmidt/httprouter"
+	"github.com/osbuild/osbuild-composer/internal/blueprint"
+	"github.com/osbuild/osbuild-composer/internal/store"
+	"github.com/osbuild/osbuild-composer/internal/target"
+)
+
+// API encapsulates RCM-specific API that is exposed over a separate TCP socket
+type API struct {
+	logger *log.Logger
+	store  *store.Store
+	router *httprouter.Router
+}
+
+// New creates new RCM API
+func New(logger *log.Logger, store *store.Store) *API {
+	api := &API{
+		logger: logger,
+		store:  store,
+		router: httprouter.New(),
+	}
+
+	api.router.RedirectTrailingSlash = false
+	api.router.RedirectFixedPath = false
+	api.router.MethodNotAllowed = http.HandlerFunc(methodNotAllowedHandler)
+	api.router.NotFound = http.HandlerFunc(notFoundHandler)
+
+	api.router.POST("/submit", api.submit)
+	api.router.GET("/status/:uuid", api.status)
+
+	return api
+}
+
+// Serve serves the RCM API over the provided listener socket
+func (api *API) Serve(listener net.Listener) error {
+	server := http.Server{Handler: api}
+
+	err := server.Serve(listener)
+	if err != nil && err != http.ErrServerClosed {
+		return err
+	}
+
+	return nil
+}
+
+// ServeHTTP logs the request, sets content-type, and forwards the request to appropriate handler
+func (api *API) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	if api.logger != nil {
+		log.Println(request.Method, request.URL.Path)
+	}
+
+	writer.Header().Set("Content-Type", "application/json; charset=utf-8")
+	api.router.ServeHTTP(writer, request)
+}
+
+func methodNotAllowedHandler(writer http.ResponseWriter, request *http.Request) {
+	writer.WriteHeader(http.StatusMethodNotAllowed)
+}
+
+func notFoundHandler(writer http.ResponseWriter, request *http.Request) {
+	writer.WriteHeader(http.StatusNotFound)
+}
+
+func (api *API) submit(writer http.ResponseWriter, request *http.Request, _ httprouter.Params) {
+	// Check some basic HTTP parameters
+	contentType := request.Header["Content-Type"]
+	if len(contentType) != 1 || contentType[0] != "application/json" {
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// JSON structure expected from the client
+	type ComposeRequest struct {
+		Distribution  string   `json:"distribution"`
+		Architectures []string `json:"architectures"`
+		Repositories  []string `json:"repositories"`
+	}
+	// Parse and verify the structure
+	var composeRequest ComposeRequest
+	decoder := json.NewDecoder(request.Body)
+	decoder.DisallowUnknownFields()
+	err := decoder.Decode(&composeRequest)
+	if err != nil || composeRequest.Distribution == "" || len(composeRequest.Architectures) == 0 || len(composeRequest.Repositories) == 0 {
+		writer.WriteHeader(http.StatusBadRequest)
```

Comment:
It would be cool to provide more information about what was wrong with the request.

Reply:
Should I return a JSON with an error message?

Reply:
I think it will be easier for RCM guys to integrate with our API if we have sane error messages. We might consider using something like this: https://github.com/go-playground/validator

Reply:
Some structured JSON errors makes sense to me. I don't know the project @ondrejbudai mentions, but on first glance something like that looks very useful (we could land the first version without though, if it is not clear which solution to pick).